encoding/ccf/decode.go

```diff
@@ -0,0 +1,1887 @@
+		return nil, err
+	}
+
+	return cadence.NewMeteredIntFromBig(
+		d.gauge,
+		common.NewCadenceIntMemoryUsage(
+			common.BigIntByteLength(bigInt),
+		),
+		func() *big.Int {
+			return bigInt
+		},
+	), nil
+}
+
+// decodeInt8 decodes int8-value as
+// language=CDDL
+// int8-value = (int .ge -128) .le 127
+func (d *Decoder) decodeInt8() (cadence.Value, error) {
+	i, err := d.dec.DecodeInt64()
+	if err != nil {
+		return nil, err
+	}
+	if i < math.MinInt8 || i > math.MaxInt8 {
+		return nil, fmt.Errorf(
+			"encoded int8-value %d is outside range of Int8 [%d, %d]",
+			i,
+			math.MaxInt8,
+			math.MaxInt8,
+		)
+	}
+	return cadence.NewMeteredInt8(d.gauge, int8(i)), nil
+}
+
+// decodeInt16 decodes int16-value as
+// language=CDDL
+// int16-value = (int .ge -32768) .le 32767
+func (d *Decoder) decodeInt16() (cadence.Value, error) {
+	i, err := d.dec.DecodeInt64()
+	if err != nil {
+		return nil, err
+	}
+	if i < math.MinInt16 || i > math.MaxInt16 {
+		return nil, fmt.Errorf(
+			"encoded int16-value %d is outside range of Int16 [%d, %d]",
+			i,
+			math.MinInt16,
+			math.MaxInt16,
+		)
+	}
+	return cadence.NewMeteredInt16(d.gauge, int16(i)), nil
+}
+
+// decodeInt32 decodes int32-value as
+// language=CDDL
+// int32-value = (int .ge -2147483648) .le 2147483647
+func (d *Decoder) decodeInt32() (cadence.Value, error) {
+	i, err := d.dec.DecodeInt64()
+	if err != nil {
+		return nil, err
+	}
+	if i < math.MinInt32 || i > math.MaxInt32 {
+		return nil, fmt.Errorf(
+			"encoded int32-value %d is outside range of Int32 [%d, %d]",
+			i,
+			math.MinInt32,
+			math.MaxInt32,
+		)
+	}
+	return cadence.NewMeteredInt32(d.gauge, int32(i)), nil
+}
+
+// decodeInt64 decodes int64-value as
+// language=CDDL
+// int64-value = (int .ge -9223372036854775808) .le 9223372036854775807
+func (d *Decoder) decodeInt64() (cadence.Value, error) {
+	i, err := d.dec.DecodeInt64()
+	if err != nil {
+		return nil, err
+	}
+	return cadence.NewMeteredInt64(d.gauge, i), nil
+}
+
+// decodeInt128 decodes int128-value as
+// language=CDDL
+// int128-value = bigint
+func (d *Decoder) decodeInt128() (cadence.Value, error) {
+	bigInt, err := d.dec.DecodeBigInt()
+	if err != nil {
+		return nil, err
+	}
+	return cadence.NewMeteredInt128FromBig(
+		d.gauge,
+		func() *big.Int {
+			return bigInt
+		},
+	)
+}
+
+// decodeInt256 decodes int256-value as
+// language=CDDL
+// int256-value = bigint
+func (d *Decoder) decodeInt256() (cadence.Value, error) {
+	bigInt, err := d.dec.DecodeBigInt()
+	if err != nil {
+		return nil, err
+	}
+	return cadence.NewMeteredInt256FromBig(
+		d.gauge,
+		func() *big.Int {
+			return bigInt
+		},
+	)
+}
+
+// decodeUInt decodes uint-value as
+// language=CDDL
+// uint-value = bigint .ge 0
+func (d *Decoder) decodeUInt() (cadence.Value, error) {
+	bigInt, err := d.dec.DecodeBigInt()
+	if err != nil {
+		return nil, err
+	}
+	if bigInt.Sign() < 0 {
+		return nil, fmt.Errorf(
+			"encoded uint-value %s is negative",
+			bigInt.String(),
+		)
```

Comment:
Here and below: The value might be very large, so formatting it as a string might allocate a large amount of memory:

```suggestion
		return nil, errors.New("encoded uint-value is negative")
```

Reply:
Good point!  Done in bc220dce48e3811da284fd94fd47e5be21c810de.

I also updated all related error messages for `big.Int`.